mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java

```diff
@@ -0,0 +1,571 @@
+    private boolean reuseContainers;
+    private boolean caseSensitive;
+
+    @Override
+    public void initialize(InputSplit split, TaskAttemptContext newContext) {
+      Configuration conf = newContext.getConfiguration();
+      // For now IcebergInputFormat does its own split planning and does not
+      // accept FileSplit instances
+      CombinedScanTask task = ((IcebergSplit) split).task;
+      this.context = newContext;
+      this.tasks = task.files().iterator();
+      this.tableSchema = SchemaParser.fromJson(conf.get(TABLE_SCHEMA));
+      String readSchemaStr = conf.get(READ_SCHEMA);
+      if (readSchemaStr != null) {
+        this.expectedSchema = SchemaParser.fromJson(readSchemaStr);
+      }
+      this.reuseContainers = conf.getBoolean(REUSE_CONTAINERS, false);
+      this.caseSensitive = conf.getBoolean(CASE_SENSITIVE, true);
+      this.inMemoryDataModel = conf.getEnum(IN_MEMORY_DATA_MODEL, InMemoryDataModel.DEFAULT);
+      this.currentIterator = open(tasks.next());
+    }
+
+    @Override
+    public boolean nextKeyValue() throws IOException {
+      while (true) {
+        if (currentIterator.hasNext()) {
+          currentRow = currentIterator.next();
+          return true;
+        } else if (tasks.hasNext()) {
+          currentCloseable.close();
+          currentIterator = open(tasks.next());
+        } else {
+          return false;
+        }
+      }
+    }
+
+    @Override
+    public Void getCurrentKey() {
+      return null;
+    }
+
+    @Override
+    public T getCurrentValue() {
+      return currentRow;
+    }
+
+    @Override
+    public float getProgress() {
+      return context.getProgress();
+    }
+
+    @Override
+    public void close() throws IOException {
+      currentCloseable.close();
+    }
+
+    private Iterator<T> open(FileScanTask currentTask) {
+      DataFile file = currentTask.file();
+      // schema of rows returned by readers
+      PartitionSpec spec = currentTask.spec();
+      Schema readSchema = expectedSchema != null ? expectedSchema : tableSchema;
+      Set<Integer> idColumns =  Sets.intersection(spec.identitySourceIds(), TypeUtil.getProjectedIds(readSchema));
+      boolean hasJoinedPartitionColumns = !idColumns.isEmpty();
+
+      if (hasJoinedPartitionColumns) {
+        Schema readDataSchema = TypeUtil.selectNot(readSchema, idColumns);
+        Schema identityPartitionSchema = TypeUtil.select(readSchema, idColumns);
+        return Iterators.transform(
+            open(currentTask, readDataSchema),
+            row -> withPartitionColumns(row, identityPartitionSchema, spec, file.partition()));
+      } else {
+        return open(currentTask, readSchema);
+      }
+    }
+
+    private Iterator<T> open(FileScanTask currentTask, Schema readSchema) {
+      DataFile file = currentTask.file();
+      // TODO should we somehow make use of FileIO to create inputFile?
+      InputFile inputFile = HadoopInputFile.fromLocation(file.path(), context.getConfiguration());
```

Comment:
In the future, we should update HadoopFileIO to use a transient `Configuration` and to keep a set of properties that it extracts from its original configuration. Then we can serialize HadoopFileIO more easily, and rebuild a working IO config on tasks.